### PR TITLE
Bug 1305507- Make sure canSaveClassifications works while entering bug number

### PR DESCRIPTION
**Pull request by KWierso**
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1883)

### CODE REVIEW

**Comment by camd** [comment]
This actually leaves you open to entering an invalid bug number (either empty or letters) and it saying True to ``canSaveClassification``.  If you are ``enteringBug`` then you may need to run through the type check for clicking enter on the bug field first before returning true to the function.  I'm not quite sure how to go about that since that's actually in the HTML of that field.  You may need to duplicate the logic.  :/

Sorry this one is a hassle.  Seems like such a simple thing.  It just has a few gotchas.

I would suggest these tests with both clicking save and ctrl+enter:

1. enter a valid bug number by hand, be able to save
2. enter text by hand, gives error when trying to save and stops the save
3. empty string for bug number, I think it still lets you save which I guess is OK.

**Comment by KWierso** [comment]
The 'enter bug number' field is marked invalid if non-numbers are entered. Can I just make the check (enteringBug && [something about that field's validity])?

**Comment by camd** [comment]
Wes--  Dang, sorry I missed the email with this question in it.  Yeah, I think you could do that.  Looks like this SO response has an idea how to use it?

http://stackoverflow.com/a/18282155/333614

-Cam

File: ui/plugins/pinboard.js [hunk 1]
```diff
@@ -84,6 +84,12 @@ treeherder.controller('PinboardCtrl', [
 
         $scope.save = function() {
             var errorFree = true;
+            if ($scope.enteringBugNumber) {
+                // we should save this for the user, as they likely
+                // just forgot to hit enter. Returns false if invalid
+                thNotify.send("Please enter a valid bug number", "danger");
+                errorFree = $scope.saveEnteredBugNumber();
+            }
```
Comment (KWierso) [inline ui/plugins/pinboard.js]: Changed this so it forces the bug number to be saved prior to checking anything else if we're in enteringBugNumber mode. This lets us validate the bug number more easily. If it isn't a valid positive integer, it warns the user about it and bails on trying to submit the classification.

File: ui/plugins/pinboard.js [hunk 2]
```diff
@@ -158,15 +159,16 @@ treeherder.controller('PinboardCtrl', [
 
         $scope.canSaveClassifications = function() {
             var thisClass = $scope.classification;
-            return $scope.hasPinnedJobs() && (thPinboard.hasRelatedBugs() && $scope.user.loggedin ||
-                   thisClass.failure_classification_id !== 4 && thisClass.failure_classification_id !== 2 ||
+            return $scope.hasPinnedJobs() && $scope.user.loggedin &&
+                   (thPinboard.hasRelatedBugs() ||
+                   (thisClass.failure_classification_id !== 4 && thisClass.failure_classification_id !== 2) ||
```
Comment (KWierso) [inline ui/plugins/pinboard.js]: This seems like it's more correctly describing valid configurations than what we had before. Could still probably be improved, though.

File: ui/plugins/pinboard.js [hunk 3]
```diff
@@ -158,15 +159,16 @@ treeherder.controller('PinboardCtrl', [
...
                    $rootScope.currentRepo.repository_group.name === "try" ||
                    $rootScope.currentRepo.repository_group.name === "project repositories" ||
                    (thisClass.failure_classification_id === 4 && thisClass.text.length > 0) ||
                    (thisClass.failure_classification_id === 2 && thisClass.text.length > 7));
         };
 
-        // Dyanmic btn/anchor title for classification save
+        // Dynamic btn/anchor title for classification save
```
Comment (KWierso) [inline ui/plugins/pinboard.js]: Bugged me too much after I noticed it to leave it alone. :)

File: ui/plugins/pinboard.js [hunk 4]
```diff
@@ -249,10 +251,16 @@ treeherder.controller('PinboardCtrl', [
...
+                    var num = parseInt($scope.newEnteredBugNumber);
+                    if (num == $scope.newEnteredBugNumber && num > 0) {
+                        $log.debug("new bug number to be saved: ",
+                                   $scope.newEnteredBugNumber);
+                        thPinboard.addBug({id:$scope.newEnteredBugNumber});
+                        $scope.toggleEnterBugNumber(false);
+                        return true;
+                    } else {
+                        return false;
+                    }
```
Comment (KWierso) [inline ui/plugins/pinboard.js]: Parse the value in the input to be an integer. If that new integer is still the same string representation AND it's above 0, let it be saved. (This basically matches the pattern being checked https://github.com/mozilla/treeherder/blob/30d3ca1b4882d61f19a237b2fd66796d8c597ed1/ui/js/directives/treeherder/main.js#L151 )
Return true in the success case and false in the failure case so we can know elsewhere how this turned out.

**Comment by KWierso** [comment]
I was unable to find the form or the form's elements in $scope (or in any of $scope's $parents up to and including $rootScope), even after adding name attributes to them. So I went a different way.

File: ui/plugins/pinboard.js (hunk 1 above)
Comment (camd) [inline ui/plugins/pinboard.js]: great!  good approach.  :)

**Pull request merged**